### PR TITLE
Adds First/Last buttons to PaginatedDataTable

### LINKED ISSUE #62191: There should be "The first page" and "The last page" button for PaginatedDataTable.
If there are thousands of table items to show, and they are sorted by Id, date of update. How could app user view the last item of current table? Click "next page" button for tens of times? That is super inconvenient.

And, the API "Goto the first page" or "Goto the last page" should be implemented and exposed to developer. For example. If the number of the items in current table could be updated dymanically, and the current page is not the first page, after the number of items is updated (like keyword searching), app user has to click "the last page" button for sometimes to go to the first page to view items.

### PR DESCRIPTION
**Pull request by jlubeck**
This PR adds the option to show First/Last buttons to the PaginatedDataTable widget.

Before:
Screen Shot 2021-01-14 at 1 52 53 PM

After:
Screen Shot 2021-01-14 at 1 52 36 PM

NOTE: Since I had to add new buttons, I had to add new Tooltips. I added the English and Spanish localizations for them, and added English in every other locale (leaving them empty was breaking the tests). Not sure what is the way to handle this, I couldn't find a mention of this in the contribution guide. Would be happy to change the PR with a better approach (hopefully you don't expect a single contributor to know all the languages 😅)

Fixes 
#62191

It's not a breaking change, so there is no migration. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing. 

**NOTE:** There is one test in  `test/material/translations_test.dart` that is failing only on my local machine, even before my changes. Seems to be passing here. 

@shihaohong looks like it's your PR: https://github.com/flutter/flutter/pull/56091 

And this is the error I'm seeing: 
```
00:09 +84: kn arb file should be properly Unicode escaped
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following FileSystemException was thrown running a test:
Cannot open file, path =
'/Users/jan/Projects/flutter/jlubeck:flutter/packages/lib/src/l10n/material_kn.arb' (OS Error: No
such file or directory, errno = 2)

When the exception was thrown, this was the stack:
#0      _File.throwIfError (dart:io/file_impl.dart:635:7)
#1      _File.openSync (dart:io/file_impl.dart:479:5)
#2      _File.readAsBytesSync (dart:io/file_impl.dart:539:18)
#3      _File.readAsStringSync (dart:io/file_impl.dart:584:18)
#4      main.<anonymous closure> (file:///Users/jan/Projects/flutter/jlubeck:flutter/packages/flutter_localizations/test/material/translations_test.dart:484:12)
#5      main.<anonymous closure> (file:///Users/jan/Projects/flutter/jlubeck:flutter/packages/flutter_localizations/test/material/translations_test.dart:478:65)
#6      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:144:29)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

The test description was:
  kn arb file should be properly Unicode escaped
════════════════════════════════════════════════════════════════════════════════════════════════════
00:09 +84 -1: kn arb file should be properly Unicode escaped [E]
  Test failed. See exception logs above.
  The test description was: kn arb file should be properly Unicode escaped

00:09 +84 -1: Some tests failed.
```

Any way I can fix my local test not passing?

### CODE REVIEW

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/paginated_data_table.dart:344]: I'm having trouble wrapping my head around this. Can you explain?

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/paginated_data_table.dart:455]: I'd break the test condition out into a one-line arrow function and replace this and the usage at 447

**Review by guidezpl — CHANGES_REQUESTED** [review]
Nice work! I'd like to see an additional test before LGTM-ing

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (guidezpl) [inline packages/flutter/test/material/paginated_data_table_test.dart:127]: I think a test to validate that buttons are disabled respectively, on the first and last pages, would be useful

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (jlubeck) [reply packages/flutter/lib/src/material/paginated_data_table.dart:344]: I actually got this algo from here: https://github.com/DataTables/DataTables/blob/de73685996c6c666e1850668c72e34d492f22150/media/js/jquery.dataTables.js#L4987 which is a well established library for DataTables which has been around for over 10 years.

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (jlubeck) [reply packages/flutter/test/material/paginated_data_table_test.dart:127]: Makes sense! Adding it right now

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (jlubeck) [reply packages/flutter/lib/src/material/paginated_data_table.dart:455]: Sounds good, done

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (jlubeck) [reply packages/flutter/test/material/paginated_data_table_test.dart:141]: Done

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

Great!

**Comment by shihaohong** [comment]
@jlubeck The instructions for adding new localized strings to Flutter can be found on the [flutter_localizations package README](https://github.com/flutter/flutter/blob/master/packages/flutter_localizations/README.md#adding-a-new-string-to-localizations). It seems like what you did looks close, but could you make sure the outcome is the same? This is because we use Google's internal translation system, which takes the English arb files to generate the strings for every other locale.

**Comment by jlubeck** [comment]
> @jlubeck The instructions for adding new localized strings to Flutter can be found on the [flutter_localizations package README](https://github.com/flutter/flutter/blob/master/packages/flutter_localizations/README.md#adding-a-new-string-to-localizations). It seems like what you did looks close, but could you make sure the outcome is the same? This is because we use Google's internal translation system, which takes the English arb files to generate the strings for every other locale.

Hey @shihaohong I just went through the README, but I'm not sure if it's outdated, or someone before me made something different. Let me explain. The way I did this was basically to copy how `nextPageTooltip` was implemented. And just make the same for `firstPageTooltip` and `lastPageTooltip`. But here is the problem. The instructions you gave me say:

`1. Add the new getter to the localizations class MaterialLocalizations, in flutter_localizations/lib/src/material_localizations.dart`

But `nextPageTooltip` is not there. It is in `flutter/lib/src/material/material_localizations.dart`

The same with number 2.

`2. Implement a default value in DefaultMaterialLocalizations in flutter_localizations/lib/src/material_localizations.dart`.

`DefaultMaterialLocalizations` is not on that file. It is also on `flutter/lib/src/material/material_localizations.dart`

Finally for `3. Add a test to test/material/localizations_test.dart that verifies that this new value is implemented.` Seeing that the README is in the `flutter_localizations` package, I was expecting to find that file in that package. But again, it was in the `flutter` package.

Now good news, running `dart dev/tools/localization/bin/gen_missing_localizations.dart` didn't give me any missing results, so that's what leads me to believe that maybe the `README` itself might be outdated?

**Comment by shihaohong** [comment]
Sorry about that, I think you're right that the README is outdated. I was actually thinking of the code comments in [packages/flutter/lib/src/material/material_localizations.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/material_localizations.dart#L14).

I wonder if running `dart dev/tools/localization/bin/gen_missing_localizations.dart` doesn't do anything because you might have manually added the English strings to all other locales already? If that's the case, everything should be fine already.

File: packages/flutter_localizations/lib/src/l10n/material_ar.arb
Comment (shihaohong) [inline packages/flutter_localizations/lib/src/l10n/material_ar.arb:24]: Shouldn't these be the English placeholders?

File: packages/flutter_localizations/lib/src/l10n/material_es.arb
Comment (shihaohong) [inline packages/flutter_localizations/lib/src/l10n/material_es.arb:13]: Shouldn't these (and all the other es variants) be the English placeholders?

Comment (shihaohong) [reply packages/flutter_localizations/lib/src/l10n/material_es.arb:13]: Edit: I just noticed you said you added the Spanish equivalents for them. Google's internal translation console will eventually provide these translations. 

We do not take community translations since 1) they will be overridden anyway by Google's translation system and 2) the engineering team has no way to check if they're correct and is not spam (for example, in the worst case, someone puts in profanity or very incorrect translations and we have no idea).

File: packages/flutter_localizations/lib/src/l10n/material_ar.arb
Comment (jlubeck) [reply packages/flutter_localizations/lib/src/l10n/material_ar.arb:24]: Based on your other comment that you didn't take community translations, yes! I actually took some of the translations from another open source project as well, but I'll turn them all back to English then

File: packages/flutter_localizations/lib/src/l10n/material_es.arb
Comment (jlubeck) [reply packages/flutter_localizations/lib/src/l10n/material_es.arb:13]: Sounds good, I'll roll them back to English. That's actually good to know that I won't have to worry about that in the future

File: packages/flutter_localizations/lib/src/l10n/material_ar.arb
Comment (jlubeck) [reply packages/flutter_localizations/lib/src/l10n/material_ar.arb]: Done

File: packages/flutter_localizations/lib/src/l10n/material_es.arb
Comment (jlubeck) [reply packages/flutter_localizations/lib/src/l10n/material_es.arb]: Done

**Comment by shihaohong** [comment]
The analyzer is failing. Don't forget to rerun `dart dev/tools/localization/bin/gen_localizations --overwrite`

**Comment by jlubeck** [comment]
@shihaohong sorry about that... Pushed again after rerunning

**Review by shihaohong — APPROVED** [review]
LGTM. No worries, and thanks for this PR! :)

**Comment by jlubeck** [comment]
@shihaohong awesome! I actually have another one in the works for this same Widget. Just wanted to get through a successful PR before sending another one. 

If you don't mind me asking, what is the next step to get this merged?

**Comment by shihaohong** [comment]
Once all the CI checks pass, and as long as the `waiting for tree to go green` tag is on, a bot will automatically merge the PR.

**Comment by jlubeck** [comment]
> Once all the CI checks pass, and as long as the `waiting for tree to go green` tag is on, a bot will automatically merge the PR.

Thanks!

**Comment by jlubeck** [comment]
Hey @shihaohong I noticed there is a Conflicting file now for some reason? Is that something I should be fixed? It says that `only those with write access to this repository can merge pull requests.`

**Comment by shihaohong** [comment]
The master branch had a change that conflicted with this PR's. I just resolved the merge conflict and it should hopefully be good to go!

**Comment by jlubeck** [comment]
@shihaohong does that merge how anything do do with that Linux flutter_plugins test failing now?

**Comment by shihaohong** [comment]
It seems unrelated to your PR and seems to be a CI issue. I've already highlighted the problem in the `hackers-infra` chat on Discord, so someone should be looking into this during the day Pacific Time (PT).

**Comment by shihaohong** [comment]
You'll need to rebase with the `master` branch, since it contains the commit that will pass the `Linux flutter_plugins` check. Sorry for the trouble!

**Comment by jlubeck** [comment]
I would be happy to do that, but it's the first time I have ever heard the word `rebase` 😅

**Comment by shihaohong** [comment]
That's fine. To make it easier, I opened a PR where I merged with master for you: https://github.com/jlubeck/flutter/pull/1. Once you merge the change to your branch, you should be good to go.

**Comment by jlubeck** [comment]
Alright, just approved it. So that's what `rebase` means? To pull newer branched from upstream?

**Comment by shihaohong** [comment]
Right! Basically, one of the commits in the `master` branch is needed for your branch to pass the checks, since without it, the CI thinks that there's something wrong with your PR, even though it's unrelated to the actual changes you're making to the codebase. So, you need to basically merge/rebase your branch with `master` to pull in those changes. 

The reason I did it for you is sometimes it can be messy if done incorrectly and require the PR to just be redone from scratch, and we already spent some time on this PR so I didn't want to go through that ordeal. Here's a quick article about it if you want to dig in and maybe try it out sometime: https://www.freecodecamp.org/news/an-introduction-to-git-merge-and-rebase-what-they-are-and-how-to-use-them-131b863785f/

**Comment by jlubeck** [comment]
Nice, thank you for that! Looks like everything is passing now. Sweet!

**Pull request merged**
